### PR TITLE
Dev 2765 page editor data layer refactor

### PR DESCRIPTION
**Pull request by nrh-cklimas**
Please fill out each section even if it's just with "N/A"

#### Plan? (if this draft/incomplete indicate what you intend to do and how)

This is the feature branch that these subtasks were merged into:

- https://news-revenue-hub.atlassian.net/browse/DEV-2864
- https://news-revenue-hub.atlassian.net/browse/DEV-2865
- https://news-revenue-hub.atlassian.net/browse/DEV-2866
- https://news-revenue-hub.atlassian.net/browse/DEV-2867

This PR is mostly for QA purposes. No other additional commits have been added.

#### What's this PR do?

Refactors the page editor logic so the majority of it occurs in hooks instead of logic embedded into components, and adds tests around those hooks.

#### Why are we doing this? How does it help us?

Gets us closer to being able to use Jest tests, improves dev speed going forward because we won't have to write code for manual API interactions going forward for new functionality.

#### How should this be manually tested? Please include detailed step-by-step instructions.

A lot has changed behind the scenes in the page editor so comprehensive QA is needed. We may also want this on staging ahead of time with multiple people, too.

Here are suggested tasks to test. For page elements, I would test with all types to make sure they all continue to work.

1. Add an element to a page and confirm that it saves properly and displays in the live contribution page.
2. Edit an element to a page and confirm it saves and displays properly.
3. Remove an element from a page and confirm it saves and displays properly.
4. Add an element to a page sidebar, confirm it saves/displays.
5. Edit a sidebar element, confirm it saves/displays.
6. Remove a sidebar element, confirm it saves/displays.
8. Check the image fields in the setup tab for regressions--able to add and remove images (and that removing one doesn't affect unrelated fields).
9. Check that the text fields in the setup tab save correctly and are reflected in the donation page.
10. Check that changing styles on a page saves/displays.
11. Check that adding a new style to a page through the Style tab works.
12. Check that creating a new page works, and that the page is editable.
13. Check that publishing a new page works and that the grab link button shows the correct URL.
14. Check that an existing published page shows as published in the top nav bar.

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

No.

#### Have automated unit tests been added? If not, why?

Yes.

#### How should this change be communicated to end users?

n/a

#### Are there any smells or added technical debt to note?

No.

#### Has this been documented? If so, where?

No.

#### What are the relevant tickets? Add a link to any relevant ones.

- https://news-revenue-hub.atlassian.net/browse/DEV-2864
- https://news-revenue-hub.atlassian.net/browse/DEV-2865
- https://news-revenue-hub.atlassian.net/browse/DEV-2866
- https://news-revenue-hub.atlassian.net/browse/DEV-2867

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

No.

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

No.

### CODE REVIEW

**Review by GuiMend — CHANGES_REQUESTED** [review]
@nrh-cklimas, I'm having issues with:
1. Adding blocks to the sidebar:
image

2. After removing a block from the "Layout" tab > saving > the item still appears there. Give it a delay of around 5s and the screen updates with the correct blocks

3. Removing blocks from sidebar. Item disapears from preview, but it still shows in the sidebar. After saving, same behavior as item 2. occurs. (After save, block shows in the preview again, wait 5s, page refreshes, removs block and closes automatically the sidebar)

These are the issues I found so far. Will continue searching

**Review by GuiMend — CHANGES_REQUESTED** [review]
Update block on sidebar showing 👍 

1. Found other issues: in the sidebar, after adding 1 block each, I added some text and uploaded an image, saved the page, tried deleting all 3 block from sidebar and this happens:
When clicking the delete button, instead of removing the block it swapped `Contribution benefits` <-> `Image` back and forth
image
image


2. When changing the orther in Layout:
image
After clicking to save, the preview shows the "old" version, and only updates back to the correct order after the save is successfull
image
image

**Review by GuiMend — APPROVED** [review]

File: spa/src/components/pageEditor/editInterface/EditInterfaceContextProvider.tsx [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+  children: ReactNode;
+}
+
+export const EditInterfaceContext = createContext<UseEditInterfaceContextResult>({
+  setElements: () => {
+    throw new Error('EditInterfaceContext must be used inside an EditInterfaceContextProvider');
+  },
+  sidebarElements: [],
+  setSidebarElements() {
+    throw new Error('EditInterfaceContext must be used inside an EditInterfaceContextProvider');
```
Comment (GuiMend) [inline spa/src/components/pageEditor/editInterface/EditInterfaceContextProvider.tsx:77]: Could add this error message into a const as it's used several times in this file

**Review by GuiMend — APPROVED** [review]

File: spa/src/components/pageEditor/editInterface/EditInterfaceContextProvider.tsx (hunk 1 above)
Comment (nrh-cklimas) [reply spa/src/components/pageEditor/editInterface/EditInterfaceContextProvider.tsx]: Updated in 27e1641e24d00488ed1f16e0dab230e52a32041d

**Review by x110dc — APPROVED** [review]

**Pull request merged**